### PR TITLE
Update to preserve firstChild in list where incomming node is the same

### PR DESCRIPTION
**Pull request by joshgillies**
Unfortunately the tests aren't exactly working - potentially due to lack of `MutationObserver` in `jsdom` https://github.com/tmpvar/jsdom/issues/639 or due to my rushed effort. 🙄 

But the test outlined illustrates the issue I was facing before this patch was developed.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -169,6 +169,68 @@ tressa.async(function (done) {
   });
 })
 .then(function () {
+  return tressa.async(function (done) {
+    tressa.log('## preserve first child where first child is the same as incomming');
```
Comment (marcoscaceres) [inline test/test.js]: Typo: s/incomming/incoming

File: test/test.js [hunk 2]
```diff
@@ -169,6 +169,68 @@ tressa.async(function (done) {
...
+      childList: true,
+      subtree: true,
+    });
+
+    var counters = [];
+
+    function trackMutations (nodes, countKey) {
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        if (nodes[i] && nodes[i].getAttribute && nodes[i].getAttribute('data-test')) {
+          (counters[nodes[i].getAttribute('data-test')] || (counters[nodes[i].getAttribute('data-test')] = { added: 0, removed: 0 }))[countKey]++;
```
Comment (marcoscaceres) [inline test/test.js]: Although fancy, could you maybe split this up a bit? Bit hard to follow. Also, nit, line length.

**Comment by joshgillies** [comment]
@marcoscaceres updated, thanks! 👍

**Comment by WebReflection** [comment]
Thaks for this PR. Unfortunately broken backend means broken TravisCI and broken builds, etc etc.

There were also minor stylish gotchas (sometimes semicolons, sometimes no semicolons) plus the logic could've been simplified a bit, which is now as such:

```js
  // accordingly with the kind of child
  // it puts its content into a parent node
  function populateNode(parent, child) {
    switch (child.nodeType) {
      case 1:
        var
          childNodes = parent.childNodes,
          length = childNodes.length
        ;
        if (0 < length && childNodes[0] === child) {
          removeNodeList(childNodes, 1);
        } else if (length !== 1) {
          resetAndPopulate(parent, child);
        }
        break;
      case 11:
        if (-1 < indexOfDiffereces(parent.childNodes, child.childNodes)) {
          resetAndPopulate(parent, child);
        }
        break;
      case 3:
        parent.textContent = child.textContent;
        break;
    }
  }
```

Last, but not least, you don't need to do this
```js
function update() {
  return hyperHTML.wire(items)`...`;
}
render`${update([])}`
```

when you can simply do this instead 😉 
```js
function update() {
  render`...`;
}
update([]);
```

If you have a bound render just use it to populate it with some content.
New test here: https://github.com/WebReflection/hyperHTML/blob/master/test/test.js#L206-L234


### TL;DR it's been merged
Basically your PR is in but I didn't just merge these changes.
Hope that's fine, and thanks again.

**Comment by joshgillies** [comment]
@WebReflection thanks for the info! It was a bit of a rushed job, and as such reflected in the code. Glad I was able to illustrate the issue in such a way that it could be included one way or the other. Thanks so much for your hard work, and continued support with `hyperHTML`! 👌
